differs/rpm_diff.go

```diff
@@ -0,0 +1,312 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package differs
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	pkgutil "github.com/GoogleCloudPlatform/container-diff/pkg/util"
+	"github.com/GoogleCloudPlatform/container-diff/util"
+	dockertar "github.com/containers/image/docker/tarfile"
+	godocker "github.com/fsouza/go-dockerclient"
+	"github.com/sirupsen/logrus"
+)
+
+type RPMAnalyzer struct {
+}
+
+// Name returns the name of the analyzer.
+func (a RPMAnalyzer) Name() string {
+	return "RPMAnalyzer"
+}
+
+// Diff compares the installed rpm packages of image1 and image2.
+func (a RPMAnalyzer) Diff(image1, image2 pkgutil.Image) (util.Result, error) {
+	diff, err := singleVersionDiff(image1, image2, a)
+	return diff, err
+}
+
+// Analyze collects information of the installed rpm packages on image.
+func (a RPMAnalyzer) Analyze(image pkgutil.Image) (util.Result, error) {
+	analysis, err := singleVersionAnalysis(image, a)
+	return analysis, err
+}
+
+// getPackages returns a map of installed rpm package on image.
+func (a RPMAnalyzer) getPackages(image pkgutil.Image) (map[string]util.PackageInfo, error) {
+	path := image.FSPath
+	packages := make(map[string]util.PackageInfo)
+	if _, err := os.Stat(path); err != nil {
+		// invalid image directory path
+		return packages, err
+	}
+
+	// try to find the rpm binary in bin/ or usr/bin/
+	rpmBinary := filepath.Join(path, "bin/rpm")
+	if _, err := os.Stat(rpmBinary); err != nil {
+		rpmBinary = filepath.Join(path, "usr/bin/rpm")
+		if _, err = os.Stat(rpmBinary); err != nil {
+			return packages, nil
+		}
+	}
+
+	return rpmDataFromContainer(image)
+}
+
+// rpmDataFromContainer runs image in a container, queries the data of
+// installed rpm packages and returns a map of packages.
+func rpmDataFromContainer(image pkgutil.Image) (map[string]util.PackageInfo, error) {
+	packages := make(map[string]util.PackageInfo)
+
+	client, err := godocker.NewClientFromEnv()
+	if err != nil {
+		return packages, err
+	}
+
+	// if image is a tar archive, we need to load it
+	imageName := image.Source
+	if pkgutil.IsTar(imageName) {
+		// NOTE: all work below is only required for multilayer tar
+		// archives generated by `docker save`.  Flat file-system
+		// archives generated by `docker create` can just be imported
+		// without potential name conflicts.  However, those are not
+		// supported yet.
+		imageName, err = generateNewArchive(image.Source)
+		if err != nil {
+			fmt.Println(err.Error())
+		}
+		defer os.Remove(imageName)
+
+		imageName, err = loadArchiveToDaemon(imageName, client)
+		if err != nil {
+			return packages, fmt.Errorf("Error loading archive: %s", err)
+		}
+		logrus.Debugf("Loaded archive as image %s", imageName)
+		// find out if we have to remove the image
+		inspImg, err := client.InspectImage(imageName)
```

Comment:
Why would there be a name conflict if we remove the tags? Shouldn't docker always give us a unique ID?

Reply:
There are two related issues.

The first one is to avoid name conflicts, so we have to `nil` the `RepoTags` field. This way we don't overwrite or rename any tags.

The second issue is that Docker won't give it a unique ID but a *hash*. If the hash of the loaded tar archive (layers) corresponds to the one of an already present image, we would remove it.

I just realized that checking for any tags via inspect isn't complete either as an already present image might have no tags. I just tested it and can confirm that the previously loaded image will falsely be removed. We really hit some limitations of the API here. I will see if we can avoid that using any other metadata, but even with timestamps we could run into race conditions.

Notice that we don't have this problem with a flat rootfs tar from `docker export`. The API and its semantics are much saner there.

Reply:
I put some more thoughts into the upper issue and experimented a lot, but I don't seem to converge to a clear preference. Potential solutions are:

(1) Don't remove the images of loaded archives at all and leave the janitorial work to the user.

(2) *Update* the `RepoTags` instead and give (try to give) *unique* tags in this namespace. As multiple instances of containerdiff may run in parallel, this requires using a filesystem lock or a reliable entropy library.

(3) Refuse to analyze tar archives in the RPMAnalyzer, and ask the user to load them as images before.

@nkubala @dlorenc What do you think?

Reply:
Good points here, this is definitely an interesting question and I'm not sure there's a perfect solution.

The most "destructive" scenario I see here is loading an archive made from exporting an old version of an image (call it `foo:latest`, with hash `abcdef`, exported to `foo.tar`), which has since been rebuilt to a new hash (`deadbeef`) and updated on the same tag in the daemon. Loading `foo.tar` into the daemon will cause the image with hash `deadbeef` to be retagged with an empty string, and image `abcdef` to be tagged as `foo:latest`, which could cause some confusion after using the tool.

Unfortunately this also means that checking for tags on the newly-loaded image won't help us know if the image needs to be left alone, because in the above scenario we just overwrote an existing image (with a tag) in the daemon, which is now an unnamed orphan image.

However, it looks to me like when doing the load, we can parse the output (like you're already doing to get the name of the newly-loaded image) and if we see something like 

`The image test:latest already exists, renaming the old one with ID sha256:<hash> to empty string`

save that hash, and then when we're done go back and retag that orphaned image with the tag it originally had, and finally remove the image we loaded in. Or at the very least, log it to the user so they know what we did. WDYT @vrothberg @dlorenc?

Reply:
@nkubala Thanks for getting into this!

I considered the renaming option as well, but I dropped it as retagging the old image to its initial name can run into conflicts as well. Some other image might have taken its place in the meantime. I see containerdiff to be useful in automation, so I think that post user interaction should be avoided.

Among other things, that's why I believe that using the `containerdiff:` namespace for our purpose is the most convenient solution for a user. It's more or less convention that a tool uses its name as a registry namespace for images. The code at the moment isn't perfect as it doesn't avoid name conflicts for the tags, but that's something we can protect us from by using a file system lock or a very good entropy.

So what I suggest doing is to use the `containerdiff:` namespace for any image created by containerdiff and use a file-system lock for the tagging.  I haven't used such locks until now in golang but https://godoc.org/github.com/nightlyone/lockfile makes a good impression on me; it also supports Windows. So what could happen is to lock `~/.containerdiff.lock`, generate a random tag until we hit one that isn't already used, load the image with the updated manifest, release the lock. Plus additional work to clean-up the lock in case it's previous owner died etc.

Reply:
Ah, yeah that makes sense. Would be nice if Docker provided some control over the image name though...hopefully you get some resolution on https://github.com/moby/moby/issues/35625. If we want to go the route of minimal interaction then I think your proposal should be fine. Want to give it a shot?

Reply:
@nkubala Sure! I will try out the idea with the file-system lock tomorrow, update the PR and drop a comment explaining the change.